### PR TITLE
SI-6539 Annotation for methods unfit for post-typer ASTs

### PR DESCRIPTION
**Pull request by retronym**
Motivated by the `.value` method in the SBT task-syntax branch,
which should only be called within the context of the argument
to a setting initialization macro.

The facility is akin to a fatal deprecation.

Review by @xeno-by @harrah

### CODE REVIEW

File: src/reflect/scala/reflect/macros/compileTimeOnly.scala [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import scala.annotation.meta._
+
+/**
+ * An annotation that designates a member should not be referred to after
+ * type checking (which includes macro expansion); it must only be used in
+ * the arguments of some other macro that will eliminate it from the AST.
+ *
+ * @param  message the error message to print during compilation if a reference remains
+ *                 after type checking
+ * @since  2.10.0
```
Comment (xeno-by) [inline src/reflect/scala/reflect/macros/compileTimeOnly.scala]: 2.10.1?

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala [hunk 2]
```diff
@@ -1376,6 +1376,11 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
         )
     }
 
+    private def checkCompileTimeOnly(sym: Symbol, pos: Position) = {
+      if (sym.isCompileTimeOnly)
+        unit.error(pos, s"reference to ${sym.fullLocationString} should not survive typechecking: ${sym.compileTimeOnlyMessage.get}")
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: Why not just show a compileTimeOnlyMessage?

File: test/files/neg/t6539.check [hunk 3]
```diff
@@ -0,0 +1,10 @@
+Test_2.scala:2: error: reference to method cto in object M should not survive typechecking: cto may only be used as an argument to m
+	M.cto // error
+          ^
```
Comment (gkossakowski) [inline test/files/neg/t6539.check]: Seems like positions for those errors are off.

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala (hunk 2 above)
Comment (harrah) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: Agree.

File: src/reflect/scala/reflect/macros/compileTimeOnly.scala (hunk 1 above)
Comment (adriaanm) [inline src/reflect/scala/reflect/macros/compileTimeOnly.scala]: indeed

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala (hunk 2 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: I've changed this, and removed the default argument for the message.

File: src/reflect/scala/reflect/macros/compileTimeOnly.scala (hunk 1 above)
Comment (retronym) [inline src/reflect/scala/reflect/macros/compileTimeOnly.scala]: done

File: test/files/neg/t6539.check (hunk 3 above)
Comment (retronym) [inline test/files/neg/t6539.check]: Oops, used tabs instead of spaces.

**Comment by harrah** [comment]
LGTM.

**Comment by xeno-by** [comment]
lgtm as well granted we agree that it's a temporary solution subject to being removed after untyped macros are implemented

**Comment by retronym** [comment]
I suspect that the untyped-macros approach might still have a problem: the compile time only macro must know all of the potential enclosing macros in which it is valid, which might stop a user of a library from writing a custom macro to replace the built-in enclosing macros. But we'll see when we get there.

**Comment by xeno-by** [comment]
Take a look here: http://www.schemeworkshop.org/2011/papers/Barzilay2011.pdf.

By default the compile-time thingie should always produce a compilation error. It's a job of an enclosing macro to wrap the block with a compile-time thingie into a binding which rebinds the compile-time thingie to something useful within the domain.

**Comment by retronym** [comment]
@xeno-by Oh right, makes total sense.

**Comment by paulp** [comment]
You know you've found your optimal audience when you can fit "thingie, thingie, thingie, something" into two lines and the response is a non-sarcastic "oh right, makes total sense."

**Comment by som-snytt** [comment]
I had the same reaction.  I would have tweeted it, but it seems that neither Henry James nor @xeno-by is tweetable.  Both are succinct as possible but >140.

**Comment by retronym** [comment]
PLS REBUILD ALL

BTW, @xeno-by I'll leave it up to you to use this on `splice`.

**Comment by xeno-by** [comment]
Oh right, thanks for reminding

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala [hunk 4]
```diff
@@ -1376,6 +1376,16 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
         )
     }
 
+    private def checkCompileTimeOnly(sym: Symbol, pos: Position) = {
+      if (sym.isCompileTimeOnly) {
+        def defaultMsg =
+          """|Reference to ${sym.fullLocationString} should not have survived past type checking,
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: is this missing the leading s for interpolaShun?

**Comment by retronym** [comment]
right you are, PLS REBUILD ALL

**Comment by paulp** [comment]
I see I'm not the only one in desperate need of that warning. Keep forgetting to implement it when I'm between distractions.

**Comment by retronym** [comment]
:cat: https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/667/console

**Comment by paulp** [comment]
PLS REBUILD ALL

**Comment by paulp** [comment]
A prayer to St. Spurious

**Comment by retronym** [comment]
> I see I'm not the only one in desperate need of that warning. Keep forgetting to implement it when I'm between distractions.

Here's another candidate warning:

``` scala
"""|abc
   |def
   |"""
```

**Comment by adriaanm** [comment]
this is not a forward compatible change, so we'll have to move the annotatoin to the compiler jar or revert this

**Comment by xeno-by** [comment]
See the last comment at https://issues.scala-lang.org/browse/SI-6539

**Pull request merged**